### PR TITLE
(PUP-10100) Redact sensitive data from arguments and failed commands

### PR DESCRIPTION
**Pull request by joshcooper**
Redact output for failed commands and exclude arguments from unqualified commands.

### CODE REVIEW

File: spec/unit/provider/exec_spec.rb [hunk 1]
```diff
@@ -31,4 +36,183 @@
...
+        # file doesn't have a valid extension and isn't a valid executable. The raised error
+        # will be Errno::EIO, which is not useful. The Windows execute code needs to raise
+        # Puppet::Util::Windows::Error so the Win32 error message is preserved.
+        pending("PUP-3561 Needs to raise a meaningful Puppet::Error") if Puppet::Util::Platform.windows?
+        expect(@logs).to include(an_object_having_attributes(level: :err, message: /'#{file}' is not executable/))
+        expect(@logs).to_not include(an_object_having_attributes(message: /#{supersecret}/))
+      end
+
+      it "redacts the arguments if the relative command cannot be resolved using the path parameter" do
+        path = tmpfile('exec')
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: The `path` variable refers to a file, so it's not a valid entry in the `path` parameter (which needs to contain directories). Agree it's not super clear what's going on there. I could create a `tmpdir` and pass that as the `path` parameter.

File: spec/unit/provider/exec_spec.rb [hunk 2]
```diff
@@ -31,4 +36,183 @@
...
+        MANIFEST
+        expect(@logs).to include(an_object_having_attributes(level: :err, message: /Could not find command '#{base}'/))
+        expect(@logs).to_not include(an_object_having_attributes(message: /#{supersecret}/))
+      end
+    end
+
+    it "redacts the command on success" do
+      apply_compiled_manifest(<<-MANIFEST)
+        exec { 'true':
+          command => Sensitive.new("ruby -e 'puts \\"#{supersecret}\\"; exit 0'"),
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: It's mostly the same except for the extra-level of escaping due to HEREDOC interpolation. Let me see if I can reuse the `ruby_success` variable from above.

File: spec/unit/provider/exec_spec.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: I simplified this so there's a file (whose unqualified basename we pass as the `command`), and a directory which is the `path`

File: spec/unit/provider/exec_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: Simplified

File: spec/unit/provider/exec_spec.rb [hunk 3]
```diff
@@ -31,4 +36,183 @@
...
+      expect(@logs).to include(an_object_having_attributes(level: :debug, message: "Executing '[redacted]'", source: /Exec\[true\]/))
+      expect(@logs).to include(an_object_having_attributes(level: :debug, message: "Executing: '[redacted]'", source: "Puppet"))
+      expect(@logs).to include(an_object_having_attributes(level: :notice, message: "executed successfully"))
+      expect(@logs).to_not include(an_object_having_attributes(message: /#{supersecret}/))
+    end
+
+    it "redacts the command on failure" do
+      apply_compiled_manifest(<<-MANIFEST)
+        exec { 'false':
+          command => Sensitive.new("ruby -e 'puts \\"supersecret\\"; exit 1'"),
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: Simplified as `echo_from_ruby_exit_1`

File: spec/unit/provider/exec_spec.rb [hunk 4]
```diff
@@ -31,4 +36,183 @@
...
+            command => Sensitive.new("ruby -e 'exit 0'"),
+            unless  => Sensitive.new("ruby -e 'puts \\"#{unlesssecret}\\"; exit 1'"),
+            path    => "#{path}",
+          }
+        MANIFEST
+        expect(@logs).to include(an_object_having_attributes(level: :debug, message: "Executing check '[redacted]'"))
+        expect(@logs).to include(an_object_having_attributes(level: :debug, message: "Executing: '[redacted]'"))
+        expect(@logs).to include(an_object_having_attributes(level: :debug, message: "[output redacted]"))
+        expect(@logs).to include(an_object_having_attributes(level: :notice, message: "executed successfully"))
+        expect(@logs).to_not include(an_object_having_attributes(message: /#{unlesssecret}/))
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: ah yeah thanks!

[the remark this answers is not in the document]
Comment (kris-bosland) [reply spec/unit/provider/exec_spec.rb]: Hmm, if this is the latest update, still seems inconsistent in the source parameter.

[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: Weird not sure what happened there. Should be fixed: https://github.com/puppetlabs/puppet/compare/29e4997955176ff0943c2f3b9e70026c3077ce8c..d3e3f1e743fc87c6c1539c9848491b55bd7dc6d0

**Comment by kris-bosland** [comment]
Yeah, still one difference between unless and onlyif, but the tests look fine otherwise.

File: spec/unit/provider/exec_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: dangit, let me fix that

[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/provider/exec_spec.rb]: Fixed it for realz

**Comment by joshcooper** [comment]
jenkins please test this on redhat7-64a,windows2016-64a

**Pull request merged**